Index: u-boot-2018.03/include/compiler.h
===================================================================
--- u-boot-2018.03.orig/include/compiler.h
+++ u-boot-2018.03/include/compiler.h
@@ -66,7 +66,7 @@ typedef uint8_t __u8;
 typedef uint16_t __u16;
 typedef uint32_t __u32;
 typedef unsigned int uint;
-typedef uint64_t __u64;
+/* typedef uint64_t __u64; */
 #ifndef linux
 typedef int __kernel_daddr_t;
 typedef unsigned int __kernel_ino_t;
