diff -uprN backports-4.19.7-1.orig/net/mac80211/Kconfig backports-4.19.7-1.new/net/mac80211/Kconfig
--- backports-4.19.7-1.orig/net/mac80211/Kconfig	2019-12-05 15:44:25.505865048 +0800
+++ backports-4.19.7-1.new/net/mac80211/Kconfig	2019-12-05 16:04:29.949915594 +0800
@@ -297,6 +297,10 @@ config MAC80211_PRIV_IOCTL
 	bool "Enable mac80211's private ioctl function, this is add by siflower"
 	depends on MAC80211
 
+config WLAN_VENDOR_RWNX
+	bool "Enable rwnx macro in mac80211, this is add by siflower"
+	depends on MAC80211
+
 config MAC80211_STA_HASH_MAX_SIZE
 	int "Station hash table maximum size" if MAC80211_DEBUG_MENU
 	default 0
