Index: hostapd-2019-08-08-ca8c2bd2/src/ap/dfs.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/dfs.c	2019-08-07 21:25:25.000000000 +0800
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/dfs.c	2020-11-11 16:54:05.355032560 +0800
@@ -812,6 +812,23 @@
 				iface->cac_started = 0;
 			}
 		}
+	} else {
+		/*
+		 * add by austin.xu:
+		 * If we switch a channel from a dfs channel which doesn't
+		 * finish dfs cac, netifd will stop this device and mac80211
+		 * will abort dfs cac. But in hostapd, we do nothing when
+		 * receive NL80211_RADAR_CAC_ABORTED cmd, and this hostapd
+		 * process will not be killed. Then netifd will create a new
+		 * hostapd process which will be conflicted with the old one
+		 * and setup will fail.
+		 * Add setup complete here and set err=1 can solute this problem.
+		 * It will finish the old process and new process can work well
+		 * when we switch a channel from a dfs channel.
+		 * If it brings more questions, I am very sorry and please
+		 * tell me and remove it.
+		 */
+		hostapd_setup_interface_complete(iface, 1);
 	}
 
 	return 0;
