Index: hostapd-2019-08-08-ca8c2bd2/src/ap/ieee802_11_ht.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/ieee802_11_ht.c	2020-11-11 10:10:05.114015297 +0800
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/ieee802_11_ht.c	2020-11-11 10:18:15.306035868 +0800
@@ -359,6 +359,29 @@
 				       HOSTAPD_MODULE_IEEE80211,
 				       HOSTAPD_LEVEL_INFO,
 				       "Switching to 20 MHz operation");
+			struct ieee80211_mgmt *resp;
+
+			/*
+			 * IEEE 802.11-REVma/D9.0 - 7.3.1.11
+			 * Return the Action frame to the source without change
+			 * except that MSB of the Category set to 1.
+			 */
+			wpa_printf(MSG_DEBUG, "IEEE 802.11: Return Action "
+						"frame back to broadcast");
+			resp = os_malloc(len);
+			if (resp == NULL)
+			      return;
+			os_memcpy(resp, mgmt, len);
+			os_memcpy(resp->da, broadcast_ether_addr, ETH_ALEN);
+			os_memcpy(resp->sa, hapd->own_addr, ETH_ALEN);
+			os_memcpy(resp->bssid, hapd->own_addr, ETH_ALEN);
+
+			if (hostapd_drv_send_mlme(hapd, resp, len, 0) < 0) {
+				wpa_printf(MSG_ERROR, "IEEE 802.11: Failed to send "
+							"Action frame");
+			}
+			os_free(resp);
+
 			iface->conf->secondary_channel = 0;
 			ieee802_11_set_beacons(iface);
 		}
