diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/acs.h hostapd-2018-05-21-62566bc2/src/ap/acs.h
--- hostapd-2018-05-21-62566bc2.orig/src/ap/acs.h	2019-11-29 10:02:49.382990184 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/acs.h	2019-11-29 10:04:32.642994517 +0800
@@ -14,7 +14,7 @@

 enum hostapd_chan_status acs_init(struct hostapd_iface *iface);
 void acs_cleanup(struct hostapd_iface *iface);
-
+void ap_dynamic_select_channel_timeout(void *eloop_ctx, void *timeout_ctx);
 #else /* CONFIG_ACS */

 static inline enum hostapd_chan_status acs_init(struct hostapd_iface *iface)
@@ -27,6 +27,11 @@ static inline void acs_cleanup(struct ho
 {
 }

+void ap_dynamic_select_channel_timeout(void *eloop_ctx, void *timeout_ctx)
+{
+	wpa_printf(MSG_ERROR, "ACS was disabled on your build, rebuild hostapd with CONFIG_ACS=y or set channel");
+	return;
+}
 #endif /* CONFIG_ACS */

 #endif /* ACS_H */
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/acs.c hostapd-2018-05-21-62566bc2/src/ap/acs.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/acs.c	2019-11-29 10:02:49.334990182 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/acs.c	2019-11-29 10:39:19.979082114 +0800
@@ -583,6 +583,24 @@ acs_find_ideal_chan(struct hostapd_iface
 		   n_chans == 2 ? 40 :
 		   80);

+	FILE *channel_file = NULL;
+	u8 acs_chan_last_chan = 0;
+	char buf[8];
+	if(iface->conf->acs_chan_dynamic_select_enable)
+	{
+		if (iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G)
+			channel_file = fopen("/var/run/24g.channel", "r");
+		else
+			channel_file = fopen("/var/run/5g.channel", "r");
+	}
+
+	if (channel_file)
+	{
+		fgets(buf, sizeof(buf), channel_file);
+		fclose(channel_file);
+		acs_chan_last_chan = atoi((char *)&buf);
+	}
+
 	for (i = 0; i < iface->current_mode->num_channels; i++) {
 		double total_weight;
 		struct acs_bias *bias, tmp_bias;
@@ -615,6 +633,16 @@ acs_find_ideal_chan(struct hostapd_iface
 			continue;
 		}

+		if(acs_chan_last_chan && acs_chan_last_chan == chan->chan)
+		{
+			//current channel(last channel) is not busy enouth to have to change it
+			if(iface->conf->acs_chan_busy_per && (chan->interference_factor * 100) < iface->conf->acs_chan_busy_per)
+			{
+				ideal_chan = chan;
+				break;
+			}
+		}
+
 		factor = 0;
 		if (acs_usable_chan(chan))
 			factor = chan->interference_factor;
@@ -678,7 +706,15 @@ acs_find_ideal_chan(struct hostapd_iface
 		factor /= total_weight;

 		bias = NULL;
-		if (iface->conf->acs_chan_bias) {
+		if(acs_chan_last_chan)
+		{
+			if(acs_chan_last_chan == chan->chan)
+			{
+				tmp_bias.channel = chan->chan;
+				tmp_bias.bias = (45 - iface->conf->acs_chan_bet_per) / 45.0;
+				bias = &tmp_bias;
+			}
+		} else if (iface->conf->acs_chan_bias) {
 			for (k = 0; k < iface->conf->num_acs_chan_bias; k++) {
 				bias = &iface->conf->acs_chan_bias[k];
 				if (bias->channel == chan->chan)
@@ -921,3 +957,19 @@ enum hostapd_chan_status acs_init(struct

 	return HOSTAPD_CHAN_ACS;
 }
+
+void ap_dynamic_select_channel_timeout(void *eloop_ctx, void *timeout_ctx)
+{
+	struct hostapd_data *hapd = eloop_ctx;
+
+	//if have sta connection,and switch_channel_even_have_sta_enable false,return
+	if(!hapd->iconf->switch_channel_even_have_sta_enable && hapd->num_sta > 0)
+	{
+		eloop_register_timeout(hapd->iconf->acs_chan_dynamic_select_period * 60, 0,
+				       ap_dynamic_select_channel_timeout, hapd, NULL);
+		return;
+	}
+
+	eloop_terminate();
+	return;
+}
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.c hostapd-2018-05-21-62566bc2/src/ap/ap_config.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.c	2019-11-29 10:02:49.334990182 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/ap_config.c	2019-11-29 10:05:27.578996823 +0800
@@ -226,6 +226,11 @@ struct hostapd_config * hostapd_config_d
 	conf->acs_ch_list.num = 0;
 #ifdef CONFIG_ACS
 	conf->acs_num_scans = 5;
+	conf->acs_chan_dynamic_select_enable = 0;
+	conf->acs_chan_dynamic_select_period = 8;
+	conf->switch_channel_even_have_sta_enable = 0;
+	conf->acs_chan_busy_per = 50;
+	conf->acs_chan_bet_per = 20;
 #endif /* CONFIG_ACS */

 	/* The third octet of the country string uses an ASCII space character
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.h hostapd-2018-05-21-62566bc2/src/ap/ap_config.h
--- hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.h	2019-11-29 10:02:49.306990181 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/ap_config.h	2019-11-29 10:07:13.447001266 +0800
@@ -817,6 +817,11 @@ struct hostapd_config {
 		double bias;
 	} *acs_chan_bias;
 	unsigned int num_acs_chan_bias;
+	int acs_chan_dynamic_select_enable;
+	unsigned int acs_chan_dynamic_select_period;
+	int switch_channel_even_have_sta_enable;
+	unsigned int acs_chan_busy_per;
+	unsigned int acs_chan_bet_per;
 #endif /* CONFIG_ACS */
 	int sta_min_dbm;
 	int prohibit_weak_sig_sta_enable;
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/hostapd.c hostapd-2018-05-21-62566bc2/src/ap/hostapd.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/hostapd.c	2019-11-29 10:02:49.382990184 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/hostapd.c	2019-11-29 10:14:34.431019772 +0800
@@ -2051,6 +2051,10 @@ dfs_offload:

 	for (j = 0; j < iface->num_bss; j++)
 		hostapd_set_own_neighbor_report(iface->bss[j]);
+	//setup of interface done,if user have dynamic channel select enable,loop a timer
+	if(hapd->iconf->acs_chan_dynamic_select_enable)
+	      eloop_register_timeout(hapd->iconf->acs_chan_dynamic_select_period * 60, 0,
+				      ap_dynamic_select_channel_timeout, hapd, NULL);

 	return 0;

diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/hw_features.c hostapd-2018-05-21-62566bc2/src/ap/hw_features.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/hw_features.c	2019-11-29 10:02:49.334990182 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/hw_features.c	2019-11-29 10:16:25.767024444 +0800
@@ -819,6 +819,18 @@ int hostapd_acs_completed(struct hostapd
 			hostapd_hw_get_freq(iface->bss[0],
 					    iface->conf->channel),
 			iface->conf->channel);
+		//write valid channel to /var/run/24g.channel /var/run/5g.channel
+		FILE *channel_file = NULL;
+		if (iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G)
+			channel_file = fopen("/var/run/24g.channel", "w+");
+		else
+			channel_file = fopen("/var/run/5g.channel", "w+");
+
+		if (channel_file)
+		{
+			fprintf(channel_file,"%d\n",iface->conf->channel);
+			fclose(channel_file);
+		}
 		break;
 	case HOSTAPD_CHAN_ACS:
 		wpa_printf(MSG_ERROR, "ACS error - reported complete, but no result available");
