diff -uprN hostapd-2018-05-21-62566bc2.orig/hostapd/config_file.c hostapd-2018-05-21-62566bc2/hostapd/config_file.c
--- hostapd-2018-05-21-62566bc2.orig/hostapd/config_file.c	2019-11-29 10:44:35.095095338 +0800
+++ hostapd-2018-05-21-62566bc2/hostapd/config_file.c	2019-11-29 10:46:43.663100734 +0800
@@ -2331,6 +2331,8 @@ static int hostapd_config_fill(struct ho
 		bss->ap_max_inactivity = atoi(pos);
 	} else if (os_strcmp(buf, "skip_inactivity_poll") == 0) {
 		bss->skip_inactivity_poll = atoi(pos);
+	} else if (os_strcmp(buf, "max_all_num_sta") == 0) {
+		conf->max_all_num_sta = atoi(pos);
 	} else if (os_strcmp(buf, "country_code") == 0) {
 		os_memcpy(conf->country, pos, 2);
 	} else if (os_strcmp(buf, "country3") == 0) {
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.c hostapd-2018-05-21-62566bc2/src/ap/ap_config.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.c	2019-11-29 10:44:35.071095337 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/ap_config.c	2019-11-29 10:47:45.283103320 +0800
@@ -189,6 +189,7 @@ struct hostapd_config * hostapd_config_d
 	hostapd_config_defaults_bss(bss);

 	conf->num_bss = 1;
+	conf->max_all_num_sta = MAX_STA_COUNT;

 	conf->beacon_int = 100;
 	conf->rts_threshold = -1; /* use driver default: 2347 */
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.h hostapd-2018-05-21-62566bc2/src/ap/ap_config.h
--- hostapd-2018-05-21-62566bc2.orig/src/ap/ap_config.h	2019-11-29 10:44:35.071095337 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/ap_config.h	2019-11-29 10:48:30.219105206 +0800
@@ -738,6 +738,7 @@ struct hostapd_config {
 	unsigned int track_sta_max_num;
 	unsigned int track_sta_max_age;

+	int max_all_num_sta;/*  to limit the linked stations number */
 	char country[3]; /* first two octets: country code as described in
 			  * ISO/IEC 3166-1. Third octet:
 			  * ' ' (ascii 32): all environments
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/beacon.c hostapd-2018-05-21-62566bc2/src/ap/beacon.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/beacon.c	2019-11-29 10:44:35.075095337 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/beacon.c	2019-11-29 11:06:58.943151734 +0800
@@ -945,7 +945,8 @@ void handle_probe_req(struct hostapd_dat
 	if (hapd->conf->no_probe_resp_if_max_sta &&
 	    is_multicast_ether_addr(mgmt->da) &&
 	    is_multicast_ether_addr(mgmt->bssid) &&
-	    hapd->num_sta >= hapd->conf->max_num_sta &&
+	    (hapd->num_sta >= hapd->conf->max_num_sta ||
+	     hapd->iface->all_num_sta >= hapd->iconf->max_all_num_sta)&&
 	    !ap_get_sta(hapd, mgmt->sa)) {
 		wpa_printf(MSG_MSGDUMP, "%s: Ignore Probe Request from " MACSTR
 			   " since no room for additional STA",
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/hostapd.h hostapd-2018-05-21-62566bc2/src/ap/hostapd.h
--- hostapd-2018-05-21-62566bc2.orig/src/ap/hostapd.h	2019-11-29 10:44:35.075095337 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/hostapd.h	2019-11-29 11:08:06.319154562 +0800
@@ -466,6 +466,8 @@ struct hostapd_iface {

 	u16 hw_flags;

+	int all_num_sta;
+
 	/* Number of associated Non-ERP stations (i.e., stations using 802.11b
 	 * in 802.11g BSS) */
 	int num_sta_non_erp;
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/sta_info.c hostapd-2018-05-21-62566bc2/src/ap/sta_info.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/sta_info.c	2019-11-29 10:44:35.075095337 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/sta_info.c	2019-11-29 11:14:26.375170511 +0800
@@ -187,6 +187,7 @@ void ap_free_sta(struct hostapd_data *ha
 			~BIT((sta->aid - 1) % 32);

 	hapd->num_sta--;
+	hapd->iface->all_num_sta--;
 	if (sta->nonerp_set) {
 		sta->nonerp_set = 0;
 		hapd->iface->num_sta_non_erp--;
@@ -701,10 +702,12 @@ struct sta_info * ap_sta_add(struct host
 		return sta;

 	wpa_printf(MSG_DEBUG, "  New STA");
-	if (hapd->num_sta >= hapd->conf->max_num_sta) {
+	if (hapd->num_sta >= hapd->conf->max_num_sta ||
+				hapd->iface->all_num_sta >= hapd->iconf->max_all_num_sta) {
 		/* FIX: might try to remove some old STAs first? */
-		wpa_printf(MSG_DEBUG, "no more room for new STAs (%d/%d)",
-			   hapd->num_sta, hapd->conf->max_num_sta);
+		wpa_printf(MSG_DEBUG, "no more room for new STAs (%d/%d), (%d/%d)",
+			   hapd->num_sta, hapd->conf->max_num_sta,
+			   hapd->iface->all_num_sta, hapd->iconf->max_all_num_sta);
 		return NULL;
 	}

@@ -736,6 +739,7 @@ struct sta_info * ap_sta_add(struct host
 	sta->next = hapd->sta_list;
 	hapd->sta_list = sta;
 	hapd->num_sta++;
+	hapd->iface->all_num_sta++;
 	ap_sta_hash_add(hapd, sta);
 	ap_sta_remove_in_other_bss(hapd, sta);
 	sta->last_seq_ctrl = WLAN_INVALID_MGMT_SEQ;
