Index: hostapd-2019-08-08-ca8c2bd2/src/ap/ap_config.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/ap_config.c	2020-11-11 17:16:13.743088307 +0800
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/ap_config.c	2020-11-11 17:20:08.495098159 +0800
@@ -168,8 +168,9 @@
 		{ aCWmin - 2, aCWmin - 1, 2, 1504 / 32, 0 };
 	const struct hostapd_tx_queue_params txq_bk =
 		{ 7, ecw2cw(aCWmin), ecw2cw(aCWmax), 0 };
+    //modify be params for make compatibility with other device when do thoughput test
 	const struct hostapd_tx_queue_params txq_be =
-		{ 3, ecw2cw(aCWmin), 4 * (ecw2cw(aCWmin) + 1) - 1, 0};
+		{ 1, 1, 3, 0};
 	const struct hostapd_tx_queue_params txq_vi =
 		{ 1, (ecw2cw(aCWmin) + 1) / 2 - 1, ecw2cw(aCWmin), 30};
 	const struct hostapd_tx_queue_params txq_vo =
