diff -uprN hostapd-2018-05-21-62566bc2.orig/hostapd/Makefile hostapd-2018-05-21-62566bc2/hostapd/Makefile
--- hostapd-2018-05-21-62566bc2.orig/hostapd/Makefile	2019-11-29 16:05:28.631903330 +0800
+++ hostapd-2018-05-21-62566bc2/hostapd/Makefile	2019-11-29 16:08:48.067911700 +0800
@@ -294,6 +294,10 @@ NEED_SHA256=y
 NEED_AES_OMAC1=y
 endif
 
+ifdef CONFIG_CUSTOM_WIFI_ONLINE
+CFLAGS += -DCONFIG_CUSTOM_WIFI_ONLINE
+endif
+
 ifdef CONFIG_IEEE80211R
 CFLAGS += -DCONFIG_IEEE80211R -DCONFIG_IEEE80211R_AP
 OBJS += ../src/ap/wpa_auth_ft.o
diff -uprN hostapd-2018-05-21-62566bc2.orig/src/ap/sta_info.c hostapd-2018-05-21-62566bc2/src/ap/sta_info.c
--- hostapd-2018-05-21-62566bc2.orig/src/ap/sta_info.c	2019-11-29 16:05:28.611903329 +0800
+++ hostapd-2018-05-21-62566bc2/src/ap/sta_info.c	2019-11-29 16:08:09.923910099 +0800
@@ -1210,6 +1210,52 @@ void ap_sta_stop_sa_query(struct hostapd
 
 #endif /* CONFIG_IEEE80211W */
 
+#ifdef CONFIG_CUSTOM_WIFI_ONLINE
+struct wifi_custom_data
+{
+	//int len;
+	u_int8_t mac[6];
+	int online;
+	char ifname[20];
+} __attribute__((packed));
+
+#define CUSTOM_WIFI_LISTEN_PORT 7892
+static broadcast_ap_sta_authorized(char *mac,char *ifname,int online)
+{
+	int sock, ret;
+	struct sockaddr_in servaddr;
+	//int ifname_len = 0;
+	struct wifi_custom_data custom_data = {0};
+
+	//ifname_len = strlen(ifname);
+	//custom_data.len = 6 + 4 + ifname_len + 1;
+	memcpy(&custom_data.mac,mac,6);
+	custom_data.online = online;
+
+	memcpy(custom_data.ifname,ifname,strlen(ifname));
+
+	//msg_len = custom_data.len + 4;
+
+	if ((sock = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)) < 0) {
+		wpa_printf(MSG_ERROR, "custom wifi socket error:");
+		return;
+	}
+	memset(&servaddr, 0, sizeof(servaddr));
+	servaddr.sin_family = AF_INET;
+	servaddr.sin_port = htons(CUSTOM_WIFI_LISTEN_PORT);
+	servaddr.sin_addr.s_addr = inet_addr("127.0.0.1");
+
+	if(sendto(sock, &custom_data, sizeof(custom_data), 0, (struct sockaddr *)&servaddr, sizeof(servaddr)) < 0)
+	{
+		wpa_printf(MSG_ERROR, "custom wifi send error");
+		goto SOCKET_ERR;
+	}
+
+SOCKET_ERR:
+	close(sock);
+	return;
+}
+#endif
 
 void ap_sta_set_authorized(struct hostapd_data *hapd, struct sta_info *sta,
 			   int authorized)
@@ -1268,6 +1314,10 @@ void ap_sta_set_authorized(struct hostap
 			wpa_msg_no_global(hapd->msg_ctx_parent, MSG_INFO,
 					  AP_STA_CONNECTED "%s%s",
 					  buf, ip_addr);
+
+#ifdef CONFIG_CUSTOM_WIFI_ONLINE
+		broadcast_ap_sta_authorized((char *)&sta->addr,(char *)&(hapd->conf->iface),1);
+#endif
 	} else {
 		wpa_msg(hapd->msg_ctx, MSG_INFO, AP_STA_DISCONNECTED "%s", buf);
 		hostapd_ubus_notify(hapd, "disassoc", sta->addr);
@@ -1276,6 +1326,10 @@ void ap_sta_set_authorized(struct hostap
 		    hapd->msg_ctx_parent != hapd->msg_ctx)
 			wpa_msg_no_global(hapd->msg_ctx_parent, MSG_INFO,
 					  AP_STA_DISCONNECTED "%s", buf);
+
+#ifdef CONFIG_CUSTOM_WIFI_ONLINE
+		broadcast_ap_sta_authorized((char *)&sta->addr,(char *)&(hapd->conf->iface),0);
+#endif
 	}
 
 #ifdef CONFIG_FST
