Index: hostapd-2019-08-08-ca8c2bd2/wpa_supplicant/events.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/wpa_supplicant/events.c	2020-11-11 16:12:06.086926837 +0800
+++ hostapd-2019-08-08-ca8c2bd2/wpa_supplicant/events.c	2020-11-11 16:16:06.626936931 +0800
@@ -3578,6 +3578,30 @@
 					     locally_generated);
 }
 
+/*
+ * Modified by nevermore, set wifi deauth/disassoc reason.
+ * Now completed by writing a non-zero value to /tmp/wds_$macaddr.
+ */
+static int set_wds_connect_status(const unsigned char *addr, int reason)
+{
+	FILE *fp;
+	char mac_buf[20];
+	char filename[32] = "/tmp/wds_reason_code";
+
+	if (addr == NULL)
+		return 0;
+
+	fp = fopen(filename, "w+");
+	if (reason == WLAN_REASON_MICHAEL_MIC_FAILURE || reason == WLAN_REASON_4WAY_HANDSHAKE_TIMEOUT ||
+			reason == WLAN_REASON_DEAUTH_LEAVING ||
+			reason == WLAN_REASON_PREV_AUTH_NOT_VALID)
+		fwrite("2\n", 1, 2, fp);
+	else
+		fwrite("1\n", 1, 2, fp);
+	fclose(fp);
+
+	return 0;
+}
 
 static void wpas_event_disassoc(struct wpa_supplicant *wpa_s,
 				struct disassoc_info *info)
@@ -3595,6 +3619,7 @@
 		ie = info->ie;
 		ie_len = info->ie_len;
 		reason_code = info->reason_code;
+		set_wds_connect_status(addr, reason_code);
 		locally_generated = info->locally_generated;
 		wpa_dbg(wpa_s, MSG_DEBUG, " * reason %u (%s)%s", reason_code,
 			reason2str(reason_code),
@@ -3650,6 +3675,7 @@
 		ie = info->ie;
 		ie_len = info->ie_len;
 		reason_code = info->reason_code;
+		set_wds_connect_status(addr, reason_code);
 		locally_generated = info->locally_generated;
 		wpa_dbg(wpa_s, MSG_DEBUG, " * reason %u (%s)%s",
 			reason_code, reason2str(reason_code),
