Index: hostapd-2019-08-08-ca8c2bd2/src/ap/acs.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/acs.c	2021-03-23 18:32:15.706906200 +0800
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/acs.c	2021-03-23 18:37:39.906915759 +0800
@@ -527,6 +527,15 @@
 	return NULL;
 }
 
+static int acs_radar_chan(const struct hostapd_channel_data *chan)
+{
+	const int allowed[] = { 52, 56, 60, 64 };
+	unsigned int i;
+	for (i = 0; i < ARRAY_SIZE(allowed); i++)
+		if (chan->chan == allowed[i])
+			return 1;
+	return 0;
+}
 
 static int is_24ghz_mode(enum hostapd_hw_mode mode)
 {
@@ -636,6 +645,9 @@
 		if (!chan_pri_allowed(chan))
 			continue;
 
+		if (acs_radar_chan(chan))
+			continue;
+
 		if (!is_in_chanlist(iface, chan))
 			continue;
 
