Index: hostapd-2019-08-08-ca8c2bd2/src/ap/ieee802_11.c
===================================================================
--- hostapd-2019-08-08-ca8c2bd2.orig/src/ap/ieee802_11.c
+++ hostapd-2019-08-08-ca8c2bd2/src/ap/ieee802_11.c
@@ -5287,10 +5287,16 @@ void ieee802_11_rx_from_unknown(struct h
 		hostapd_drv_sta_disassoc(
 			hapd, src,
 			WLAN_REASON_CLASS3_FRAME_FROM_NONASSOC_STA);
-	else
-		hostapd_drv_sta_deauth(
-			hapd, src,
-			WLAN_REASON_CLASS3_FRAME_FROM_NONASSOC_STA);
+	else if (sta == NULL) {
+		wpa_printf(MSG_DEBUG, "An empty data frame from an unassociated STA that exists in a linked list"
+		   MACSTR, MAC2STR(src));
+		return;
+	}else {
+		wpa_printf(MSG_DEBUG, "Empty data frame from unassociated STA"
+		   MACSTR, MAC2STR(src));
+		ap_free_sta(hapd, sta);
+		return;
+	}
 }
 
 
